src/azure-cli-core/azure/cli/core/azclierror.py

```diff
@@ -4,85 +4,178 @@
-        if self.raw_exception:
-            telemetry.set_exception(self.raw_exception, '')
+        telemetry.set_error_type(self.__class__.__name__)
+# endregion
+
+
+# region: Second Layer
+# Main categories of the AzureCLI error types, used for Telemetry alalysis
+# DO NOT raise the error classes here directly in your codes.
+class UserFault(AzCLIError):
+    """ Users should be responsible for the errors. """
+    def send_telemetry(self):
+        super().send_telemetry()
+        telemetry.set_user_fault(self.error_msg)
+
+
+class ServiceError(AzCLIError):
+    """ Azure Services should be responsible for the errors.  """
+    def send_telemetry(self):
+        super().send_telemetry()
+        telemetry.set_failure(self.error_msg)
+
+
+class ClientError(AzCLIError):
+    """ Azure CLI should be responsible for the errors. """
+    def send_telemetry(self):
+        super().send_telemetry()
+        telemetry.set_failure(self.error_msg)
+        if self.exception_trace:
+            telemetry.set_exception(self.exception_trace, '')
+# endregion
+
+
+# region: Third Layer
+# Sub-categories of the AzureCLI error types, shown to users
+# Raise the error classes here in your codes
+# Command related error types
+class CommandNotFoundError(UserFault):
+    """ Command is misspelled or not recognized by AzureCLI. """
+    pass
+
+
+# Argument related error types
+class UnrecognizedArgumentError(UserFault):
+    """ Argument is misspelled or not recognized by AzureCLI. """
+    pass
+
+
+class RequiredArgumentMissingError(UserFault):
+    """ Required argument is not specified. """
+    pass
+
+
+class ConflictArgumentError(UserFault):
+    """ Arguments can not be specfied together. """
+    pass
+
+
+class InvalidArgumentValueError(UserFault):
+    """ Argument value is not valid. """
+    pass
+
+
+class ArgumentParseError(UserFault):
+    """ Fallback of the argument parsing related errors.
+    Use this error only when the argument related errors
+    can not be classified into the above error types. """
+    pass
+
+
+# Response related error types
+class BadRequestError(UserFault):
+    """ Bad request from client: 400 error """
+    pass
+
+
+class UnauthorizedError(UserFault):
+    """ Unauthorized request: 401 error """
+
+
+class ForbiddenError(UserFault):
+    """ Service refuse to response: 403 error """
+
+
+class ResourceNotFoundError(UserFault):
+    """ Can not find Azure resources: 404 error """
+    pass
+
+
+class AzureInternalError(ServiceError):
+    """ Azure service internal error: 5xx error """
+    pass
+
+
+class AzureResponseError(UserFault):
+    """ Fallback of the response related errors.
+    Use this error only when the response related errors
+    can not be classified into the above error types. """
+
+
+# Request related error types
+class AzureConnectionError(UserFault):
+    """ Connection issues like connection timeout, aborted or broken. """
+    pass
+
+
+class ClientRequestError(UserFault):
```

Comment:
Little ambiguous, does this mean 400 badrequest? or userFault? or Client CLI error?